Automattic/woocommerce-payments
# Enable customers to complete checkout for subscription products with a free trial

Fixes #1979 

#### Changes proposed in this Pull Request
> As noted in this [Github thread](https://github.com/Automattic/woocommerce-payments/issues/1828#issuecomment-850055235), it is not possible to checkout with UPE with a 100% discount. This will be an issue for any purchases with 0 cost. Another example is a subscription with a free trial, where the payment method is needed to be captured.

# Free Trial Subscriptions: The Final Frontier

The functionality to support saving payment methods and adding support for checkouts involving subscription products has been implemented in the UPE gateway. However, in the case of subscription products with a free trial, there is no initial payment, but we still need to collect payment information, so that the payment method can be saved in order to complete payment once the free trial period has ended. In order to facilitate this we need to load the UPE element with a Setup Intent instead of a Payment Intent, confirm payment with the Stripe UPE element, and then complete the order.

The changes in this PR adds functionality to check whether payment is required at checkout and return a Setup Intent instead of a Payment Intent to the client AJAX request if no payment is necessary. Similarly, if no payment is required, we use the Stripe API function `confirmSetup` instead of `confirmPayment` to complete the transaction. Since for non-subscription products, if there is no payment required at checkout, the entire payment section is removed at checkout, a subscription product with no initial payment is the only case where the payment client JS will be present, yet not payment will be required. Additionally, in my testing, when I tried to pay for a subscription product with a free trial using the Pay For Order page, I was redirected to the checkout page instead to complete the order. Consequently, this code, I believe, will only affect checkouts. 

The main difference between the structure of a Setup Intent and a Payment Intent, is that the former will lack any accompanying charge objects, since there is no charge processed at the time of the transaction. This means that we do not have the complete [payment method details](https://stripe.com/docs/api/charges/object#charge_object-payment_method_details) usually captured within the charge object for the payment method used. Currently, the only payment method within the UPE that functionally supports saving payments is the credit/debit card payment method. Since we do not have access to the payment method details at checkout, without making a separate request to retrieve the full payment method details (which I have declined to do), we are not aware of the brand (e.g. visa, mastercard, etc.) or type of funding (i.e credit or debit). For this reason, the listed payment method shown on the order complete page falls back to the default card value: "Credit card / debit card (WooCommerce Payments)". This will not be an issue for Sofort, since all orders will use the default payment method title. 

![Order received page for free trial subscription product](https://cdn-std.droplr.net/files/acc_1104206/NaNKOx)
_Order received page for free trial subscription product_




#### Testing instructions





1. Ensure you have pulled the changes in this PR, the latest version of the Payments Server, and that you have the WC Subscriptions plugin enabled in your site. Also, please enable the UPE gateway using the WCPay Dev Plugin or simply set `_wcpay_feature_upe` to a value of 1 in your `wp_options` table.
2. Create a new simple subscription product with a free trial.
![Sample product details](https://cdn-std.droplr.net/files/acc_1104206/Ek8ntx)
_Sample product details for simple subscription product_

3. Add this product to your cart and proceed to checkout. You will currently only be able to use the card gateway to checkout, since this is the only gateway that currently supports saved payments.
4. Pay for your order and complete the checkout. 
5. View your orders (**WooCommerce > Orders**) and confirm that the order exists with a status of `Processing` and a total of $0.00 (or whichever currency you have set). The order should have a related Subscription order with a status of `Active` and a `Trial End` and `Next Payment` date that are both equal to each other.
6. Take a deep breath and think about something that makes you smile.

-------------------
